tests/RegistryTests.cpp

```diff
@@ -2971,3 +2971,244 @@ TEST_CASE("BasicRegistryTests::cotaskmem_array-bytes", "[registry]]")
     }
 }
 #endif // #if defined(__WIL_OBJBASE_H_)
+
+#if defined(_VECTOR_) && defined(WIL_ENABLE_EXCEPTIONS)
+TEST_CASE("BasicRegistryTests::enumerate_values", "[registry]]")
+{
+    const auto deleteHr = HRESULT_FROM_WIN32(::RegDeleteTreeW(HKEY_CURRENT_USER, testSubkey));
+    if (deleteHr != HRESULT_FROM_WIN32(ERROR_FILE_NOT_FOUND))
+    {
+        REQUIRE_SUCCEEDED(deleteHr);
+    }
+
+    SECTION("enumerate_values with no values")
+    {
+        const wil::unique_hkey hkey{wil::reg::create_unique_key(HKEY_CURRENT_USER, testSubkey)};
+        const auto value_enum = wil::reg::enumerate_values(hkey.get());
+
+        REQUIRE(value_enum.begin() == value_enum.end());
+    }
+
+    SECTION("enumerate_values with one value - manual iterator usage")
+    {
+        wil::unique_hkey hkey{wil::reg::create_unique_key(HKEY_CURRENT_USER, testSubkey, wil::reg::key_access::readwrite)};
+        wil::reg::set_value(hkey.get(), L"first_value", 0);
+
+        const auto value_enum = wil::reg::enumerate_values(hkey.get());
+        REQUIRE(value_enum.begin() != value_enum.end());
+        auto value_iterator = value_enum.begin();
+        REQUIRE(std::wstring(*value_iterator) == L"first_value");
+        ++value_iterator;
+        REQUIRE(value_iterator == value_enum.end());
+
+        value_iterator = value_enum.begin();
+        REQUIRE(std::wstring(*value_iterator) == L"first_value");
+        value_iterator += 1;
+        REQUIRE(value_iterator == value_enum.end());
+
+        const auto uniquevalue_enum = wil::reg::enumerate_values(wistd::move(hkey));
+        REQUIRE(uniquevalue_enum.begin() != uniquevalue_enum.end());
+        value_iterator = uniquevalue_enum.begin();
+        REQUIRE(std::wstring(*value_iterator) == L"first_value");
+        ++value_iterator;
+        REQUIRE(value_iterator == uniquevalue_enum.end());
+
+        value_iterator = uniquevalue_enum.begin();
+        REQUIRE(std::wstring(*value_iterator) == L"first_value");
+        value_iterator += 1;
+        REQUIRE(value_iterator == uniquevalue_enum.end());
+    }
+
+    SECTION("enumerate_values with one value - std::for_each usage")
+    {
+        wil::unique_hkey hkey{wil::reg::create_unique_key(HKEY_CURRENT_USER, testSubkey, wil::reg::key_access::readwrite)};
+        wil::reg::set_value(hkey.get(), L"first_value", 0);
+
+        uint32_t count = 0;
+        auto enumerator = wil::reg::enumerate_values(hkey.get());
+        std::for_each(enumerator.begin(), enumerator.end(), [&](PCWSTR key_name) {

```

Comment:
`value_name` or just `name` as this is a "value" case. here and below

Reply:
yep

Reply:
I think I fixed all instances.